### DISCUSSION TITLE
how to hot load a new version rabbitmq source code

### DISCUSSION THREAD

**Discussion by hzhang7** [rabbitmq/rabbitmq-server#9126] [Questions] [answered]
can I hot load a new version rabbitmq source code when I change some code?
I have tried rabbitmqctl eval "code:load_file(MODULE)." but after i load_file, rabbitmq terminated.

**Comment by lukebakken** [comment]
In general this is not supported. @mkuratczyk may have some tricks up his sleeves, though.

**Comment by mkuratczyk** [comment, marked as answer]
With the caveat that this is strictly for development purposes, yes, you can, although in some cases it's not sufficient (eg. if you change initialization code of a gen_server process, you will need to restart it after loading the new code). There are multiple ways to achieve that - manually or with erlang_ls: https://github.com/rabbitmq/contribute/wiki/Bazel#hot-code-reloading.

Manually, it's mostly about:
1. Starting the broker in a dev mode (`-c dbg` with bazel, I don't remember how to do that with `make`)
2. `c(module).` from the shell

I've never seen that terminate RabbitMQ.

With erlang_ls, you can have it reload the module on file save in your editor (it's also possible to set it up with just your editor, ultimately it's the same `c(module).` call).

**Reply by hzhang7** [reply]
thx a lot, I will try the way you mentioned above.
